### PR TITLE
HC-58: Fix Problems Scheduling Unlimited Resources

### PR DESCRIPTION
**Pull request by MiyaNoctem**
## Overview
When adding unlimited resources to booking, choosing the resource you want to add using the select combo box should load the options configured for that resource (eg. Choosing 'Hot Beverages' loads options 'Tea' and 'Coffee'). This was not happening, on some cases loading options for a different resource instead, while on others the options simply never loaded.

## Before
Loading of options is done via a series of chained API calls, from Resource, to Resource Set, to Resource Set Options. A bug injected on CiviCRM v4.7.19 on PR https://github.com/civicrm/civicrm-core/pull/10033, caused chained API calls using '$value.' to chain values to be interpreted as operators with syntax: array('OP' => $val) instead. This only happened if the first parameter for the chained API call uses '$value.'. The bug has been fixed on CiviCRM core and is set to be released on v4.7.28.

## After
Fixed by implementing civicrm_apiWrappers hook to preprocess parameters array to make sure the first parameter of a chained API call is not a chained value. This way, the calls should work both on versions with the bug and versions without it.

## Comments
Line that injects the bug: https://github.com/civicrm/civicrm-core/blob/4.7.19/api/api.php#L136
Fix done on CiviCRM: https://github.com/civicrm/civicrm-core/blob/4.7.28-rc/api/api.php#L132-L135

### CODE REVIEW

File: CRM/Booking/APIWrapper.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  private function fixParametersArray(&$params) {
+    $chainedValues = array();
+
+    foreach ($params as $parameter => &$value) {
+      if (stripos($parameter, 'api.') === 0 && is_array($value)) {
+        $allChainedValues = true;
+
+        foreach ($value as $chainedParameter => $chainedValue) {
+          if (stripos($value, '$value.') !== 0) {
+            $allChainedValues = false;
```
Comment (mickadoo) [inline CRM/Booking/APIWrapper.php]: I think with Drupal coding standards booleans should be upper case.

File: CRM/Booking/APIWrapper.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if ($allChainedValues) {
+          $value['sequential'] = 0;
+        }
+
+        $this->fixParametersArray($value);
+      }
+      elseif (stripos($value, '$value.') === 0) {
+        unset($params[$parameter]);
```
Comment (mickadoo) [inline CRM/Booking/APIWrapper.php]: I wasn't so sure, but is the purpose of this to just shift the parameter with key `$value.id` to the end of the array?

File: CRM/Booking/APIWrapper.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   * @param array $params
+   */
+  private function fixParametersArray(&$params) {
+    $chainedValues = array();
+
+    foreach ($params as $parameter => &$value) {
+      if (stripos($parameter, 'api.') === 0 && is_array($value)) {
+        $allChainedValues = true;
+
+        foreach ($value as $chainedParameter => $chainedValue) {
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/Booking/APIWrapper.php]: Not possible, as per the `is_array($value)` on lie 41.

File: CRM/Booking/APIWrapper.php (hunk 2 above)
Comment (MiyaNoctem) [reply CRM/Booking/APIWrapper.php]: Kind of, the idea is to shift all parameters with value starting with '$value.' to the end of the array.

**Pull request merged**